runtime/runtime/src/actions.rs

```diff
@@ -0,0 +1,389 @@
+            .ok_or_else(|| format!("cannot find contract code for account {}", account_id.clone()))
+    };
+    wasm::cache::get_code_with_cache(code_hash, code)
+}
+
+pub(crate) fn action_function_call(
+    state_update: &mut TrieUpdate,
+    apply_state: &ApplyState,
+    account: &mut Option<Account>,
+    receipt: &Receipt,
+    action_receipt: &ActionReceipt,
+    input_data: &Vec<Option<Vec<u8>>>,
+    result: &mut ActionResult,
+    account_id: &AccountId,
+    function_call: &FunctionCallAction,
+    action_hash: &CryptoHash,
+    config: &RuntimeConfig,
+    ethash_provider: Arc<Mutex<EthashProvider>>,
+) {
+    let account = account.as_mut().unwrap();
+    let code = match get_code_with_cache(state_update, account_id, &account) {
+        Ok(code) => code,
+        Err(e) => {
+            result.result = Err(e.into());
+            return;
+        }
+    };
+    let mut runtime_ext = RuntimeExt::new(
+        state_update,
+        account_id,
+        ethash_provider,
+        &action_receipt.signer_id,
+        &action_receipt.signer_public_key,
+        action_receipt.gas_price,
+        action_hash,
+    );
+    account.amount += function_call.deposit;
+    let start_gas = function_call.gas;
+    let mut wasm_res = match executor::execute(
+        &code,
+        &function_call.method_name.as_bytes().to_vec(),
+        &function_call.args,
+        &input_data,
+        &mut runtime_ext,
+        &config.wasm_config,
+        &RuntimeContext::new(
+            account.amount,
+            (function_call.gas as Balance) * action_receipt.gas_price,
+            &receipt.predecessor_id,
+            account_id,
+            account.storage_usage,
+            apply_state.block_index,
+            receipt.receipt_id.as_ref().to_vec(),
+            false,
+            &action_receipt.signer_id,
+            &action_receipt.signer_public_key,
+        ),
+    ) {
+        Ok(exo) => exo,
+        Err(e) => {
+            result.result =
+                Err(format!("wasm async call preparation failed with error: {:?}", e).into());
+            return;
+        }
+    };
+    result.logs.append(&mut wasm_res.logs);
+    account.storage_usage = wasm_res.storage_usage;
+    let remaining_gas = (wasm_res.liquid_balance / action_receipt.gas_price) as u64;
+    let gas_used = if start_gas > remaining_gas { start_gas - remaining_gas } else { 0 };
+    result.gas_burnt += gas_used;
+    result.gas_used += gas_used;
+    result.result = wasm_res
+        .return_data
+        .map_err(|e| format!("wasm async call execution failed with error: {:?}", e).into());
+    result.new_receipts.append(&mut runtime_ext.into_receipts(account_id));
+}
+
+pub(crate) fn action_stake(
+    account: &mut Option<Account>,
+    result: &mut ActionResult,
+    account_id: &AccountId,
+    stake: &StakeAction,
+) {
+    let mut account = account.as_mut().unwrap();
+    let increment = if stake.stake > account.staked { stake.stake - account.staked } else { 0 };
```

Comment:
Is it how we stake? If I staked already X and I want to increase my stake to X+Y then I issue a staking transaction with value X+Y instead of Y, correct?

Reply:
Yes